llvm/lib/CodeGen/MachineTraceMetrics.cpp

```diff
@@ -761,6 +762,64 @@ static void updatePhysDepsDownwards(const MachineInstr *UseMI,
   }
 }
 
+/// Estimates the number of cycles elapsed between DefMI and UseMI if they're
+/// non-null and in the same BasicBlock. Returns std::nullopt when UseMI is in a
+/// different MBB than DefMI.
+static std::optional<unsigned>
+estimateDefUseCycles(const TargetSchedModel &Sched, const MachineInstr *DefMI,
+                     const MachineInstr *UseMI) {
+  if (!DefMI || !UseMI || DefMI == UseMI)
+    return 0;
+  const MachineBasicBlock *ParentBB = DefMI->getParent();
+  if (ParentBB != UseMI->getParent())
+    return std::nullopt;
+
+  const auto DefIt =
+      llvm::find_if(ParentBB->instrs(),
+                    [DefMI](const MachineInstr &MI) { return DefMI == &MI; });
+  const auto UseIt =
+      llvm::find_if(ParentBB->instrs(),
+                    [UseMI](const MachineInstr &MI) { return UseMI == &MI; });
+
+  unsigned NumMicroOps = 0;
+  for (auto It = DefIt; It != UseIt; ++It) {
+    // In cases where the UseMI is a PHI at the beginning of the MBB, compute
+    // MicroOps until the end of the MBB.
+    if (It.isEnd())
+      break;
+
+    NumMicroOps += Sched.getNumMicroOps(&*It);
+  }
+  return NumMicroOps / Sched.getIssueWidth();
+}
+
+/// Wraps Sched.computeOperandLatency, accounting for the case when
+/// InstrSchedModel and InstrItineraries are not available: in this case,
+/// Sched.computeOperandLatency returns DefaultDefLatency, which is a very rough
+/// approximate; to improve this approximate, offset it by the approximate
+/// cycles elapsed from DefMI to UseMI (since the MIs could be re-ordered by the
+/// scheduler, and we don't have this information, this cannot be known
+/// exactly). When scheduling information is available,
+/// Sched.computeOperandLatency returns a much better estimate (especially if
+/// UseMI is non-null), so we just return that.
+static unsigned computeOperandLatency(const TargetSchedModel &Sched,
+                                      const MachineInstr *DefMI,
+                                      unsigned DefOperIdx,
+                                      const MachineInstr *UseMI,
+                                      unsigned UseOperIdx) {
+  assert(DefMI && "Non-null DefMI expected");
+  if (!Sched.hasInstrSchedModel() && !Sched.hasInstrItineraries()) {
+    unsigned DefaultDefLatency = Sched.getInstrInfo()->defaultDefLatency(
```

Comment:
Would it be better to use `DefaultDefLatency` when `DefMI.mayLoad()` or `isHighLatencyDef()` since those are two cases where the cycles elapsed between instructions are likely not a good representation (See TargetInstrInfo::defaultDefLatency).

Reply:
Hm, could you explain why you think the cycles elapsed between the two instructions isn't a good representation in these two cases? Unless I'm very much mistaken, I think most of the benefits of this patch are seen when the default def latency is non-trivial (not equal to 1): see the vector test changes.

Reply:
Example 1:
```
defmi = ...
load = ...
usemi = ...
```
The assumption you make in this patch is that DefaultDefLatency is a bad approximate. I'm not sure this is true for loads since we have a value that intends to describe how loads behave by default.

Example 2:
```
defmi = ...
instruction with high latency = ...
usemi = ...
```
I don't think assuming a value of 1 is better than the value TargetInstrInfo::defaultDefLatency uses when isHighLatencyDef.

Reply:
Sorry, I'm still trying to understand this. We have:

```cpp
unsigned TargetInstrInfo::defaultDefLatency(const MCSchedModel &SchedModel,
                                            const MachineInstr &DefMI) const {
  if (DefMI.isTransient())
    return 0;
  if (DefMI.mayLoad())
    return SchedModel.LoadLatency;
  if (isHighLatencyDef(DefMI.getOpcode()))
    return SchedModel.HighLatency;
  return 1;
}
```

This is an implementation detail of `defaultDefLatency`: my underlying assumption is that `defaultDefLatency` will always return a a first-approximate of the latency of a DefMI, without ever bothering about how it is used, and independent of its actual implementation. Whether it is transient, a load, a high-latency def, or otherwise, the function should return something that doesn't leak into its implementation detail, no?

Reply:
Hi, I had another look at the examples in your comment, and I think what you're saying is the point we already discussed: will the load or high-latency-def not be _dispatched_ in one cycle? We're assuming multi-pipeline here.

Reply:
Take this example:

```
defmi = ...
load = ...
usemi = ...
```

I think with your patch you are estimating 2 cycles between `defmi` and `usemi` with issue width = 1 (1 cycle for defmi and one cycle for load). My question is whether this is a good estimation since loads typically have different latency than the normal case. I gave it some more thought though and if `usemi` doesn't use `load` result then this does not matter, and what you have looks good.

If it does use `load` though, then we'd be calling `computeOperandLatency` between `load` and `usemi`. In that case though, should we be taking `LoadLatency` into account?

Reply:
Thanks for thinking about this case. However, I'm still confused: wouldn't `computeOperandLatency` be called with `DefMI = load` and `UseMI = usemi` when `usemi` uses the result of the `load`, and isn't this the correct estimation? I'm not sure if the `load` will be marked as a def though: it might require a MachineMemorySSA unless the case is already accounted for?

Reply:
> wouldn't computeOperandLatency be called with DefMI = load and UseMI = usemi when usemi uses the result of the load

Initially, I missed this point. But in my last response I realized I missed that. That's what I meant when I said:

> If it does use load though, then we'd be calling computeOperandLatency between load and usemi.

> And isn't this the correct estimation

We would get `DefaultDefLatency` which would account for it being a load and then subtract the number of cycles elapsed in between via `estimateDefUseCycles`. So yeah I think it works as expected.

I will resolve this conversation.

Reply:
Small caveat. I think the dependencies between loads and stores aren't captured correctly by MTM; I'd investigated this point earlier, and this is due to an infrastructural deficiency: we don't have a MachineMemorySSA.

Reply:
Even in the simple case?
```
defmi = ...
a = load
usemi = a
```
We know here for example that `a` is 100% used by `usemi`.

Reply:
I think this case should be handled just fine. I think this is the problematic case:

```
store %mem, %val.a
%val.b = load %mem
```

Sorry, it's a bit tangential to the patch itself.